vyper/parser/pre_parser.py

```diff
@@ -37,20 +39,45 @@ def parse_version_pragma(version_str):
 # Minor pre-parser checks.
 def pre_parse(code):
     result = []
+    line_ofst = 0
 
     try:
         g = tokenize(io.BytesIO(code.encode('utf-8')).readline)
+
         for token in g:
-            # Alias contract definition to class definition.
+            start = (token.start[0] + line_ofst, token.start[1])
+            end = (token.end[0] + line_ofst, token.end[1])
+            toks = [ token._replace(start=start, end=end) ]
+            line = token.line
+
             if token.type == COMMENT and "@version" in token.string:
                 parse_version_pragma(token.string[1:])
+
+            # Alias contract definition to class definition.
             if (token.type, token.string, token.start[1]) == (NAME, "contract", 0):
-                token = TokenInfo(token.type, "class", token.start, token.end, token.line)
+                line_ofst += 1
+                toks = [ TokenInfo(NAME, "@contract", start,start,line),
+                         TokenInfo(NEWLINE, "\n", start,start,line) ]
+                start = (start[0] + 1, start[1])
+                end = (end[0] + 1, end[1])
+                toks.append(
+                         TokenInfo(token.type, "class", start,end,line))
+
+            # Alias struct definition to class definition.
+            if (token.type, token.string, token.start[1]) == (NAME, "struct", 0):
+                line_ofst += 1
+                toks = [ TokenInfo(NAME, "@struct", start,start,line),
+                         TokenInfo(NEWLINE, "\n", start,start,line) ]
+                start = (start[0] + 1, start[1])
+                end = (end[0] + 1, end[1])
+                toks.append(
+                         TokenInfo(token.type, "class", start,end,line))
```

Comment:
This works, I was considering using an internal structure to just keep track of classes, do think that would be cleaner?

l. 71-74, l.61-64 are the same so you can just unindent one and remove the other.

Reply:
@charles-cooper so the idea was we just return the information for the "pre-parser" to the rest of the code in parser.parse(). Something like: